droidmediacamera2.cpp

```diff
@@ -1602,6 +1623,20 @@ bool droid_media_camera_set_parameters(DroidMediaCamera *camera, const char *par
         update_request(camera, camera->m_video_request, param_map);
     }
 
+    ALOGI("update_request preview");
+    if (!camera->m_preview_request) {
+        ALOGI("Not even preview, copy to try again later");
+        // In addition to holding the first sent params until preview and other requests are available
+        // it seems exposure compensation is only sent the first time, and there are more than one calls before
+        // the preview and other requests are available, but only the first has this parameter from gstreamer.
+        // Solution: propagate the initial exposure compensation to all subsequent retries before the preview and other requests are 'on'.
+        std::string exposure_compensation = camera->m_param_map["exposure-compensation"];
+        camera->m_param_map = param_map;
+        if (exposure_compensation.size()) {
```

Comment:
```suggestion
        if (param_map.count("exposure-compensation") == 0 && exposure_compensation.size()) {
```
from my understanding there is no longer a way to actually change the value with your code?

Reply:
First, the code: the `[]` operator returns a default constructed string if there's none, that's why I check for size. It also creates an entry, so the `count` may fail next.
Indeed, checking for its' presence with count would be better.

Secondly: The quick hack is around the fact that `jolla-camera` (or `gst-droid` itself) does not initialize the camera with the _previous_ exposure compensation setting but ONCE. This change propagates that one time value to all the params at initialization until there's an actual request (be it preview, image or video request) is set - so the code only runs when none of the above are set.
Then, the usual get params / set params should work. So the `jolla-camera` can now set other values and store other values to use for the next session. My hunch is that the lack of setting the exposure-compensation at the beginning (because not even preview is on) triggers a read of the default compensation and we're left with that - but I am open to being corrected!

Thanks for your comment! This is just a "here, I got it working on my device" PR and it is a HACK. Gstreamer should be aware of this initialization process (maybe?) and act properly with camera2.